find_email_addresses.py

```diff
@@ -0,0 +1,98 @@
+'''
+Script for scraping all discoverable websites on a domain for email addresses
+Usage: python find_email_addresses.py <domain>
+
+How it works:
+- Given a domain, it searches the page for emails and add it to a list of emails
+- Then, it searches the page of links within the href attribution within <a> tags
+- It cleans up the links, adds it to a queue, and repeats from step one with every link
+- Once the queue is empty, all emails found will outputted
+
+Created by Vincent Tran, 10/7/2016
+'''
+
+import sys
+import re
+import mechanize
+from collections import deque
+from bs4 		 import BeautifulSoup
+
+''' 
+Takes a string text to search for emails
+Returns a list of all emails found
+'''
+def grabEmails(text):
+	return re.findall(regex, text)
+
+'''
+Takes a BeautifulSoup object (parsed html) and a string url domain
+Returns a list of links that are within the given domain
+'''
+def grabLinks(parsed_text, domain):
+	links = []
+	
+	# Find links by searching for <href> inside of <a>
+	for a in parsed_text.find_all('a'):
+		link = a.get('href')
+
+		if 'http' not in link:		# Account for relative links
+			link = domain + link
+		if domain in link:		# Only accept links within given domain
+			links.append(link)
+
+	return links
+
+'''
+Takes a string domain and a compiled regex
+Returns a list of all emails found
+'''
+def scrapeEmails(domain, regex):
+	urlQueue 	= deque()					# Queue to track urls to visit
+	urlsVisited = set()						# Set to track urls already visited
+	emails 		= set()
+	urlQueue.append(domain)
+	browser 	= mechanize.Browser()
+
+	while len(urlQueue):
+		url = urlQueue.popleft()
+		if url in urlsVisited:				# Already visited, skip
+			continue
+
+		try:
+			page 		= browser.open(url)
+			source_code = page.read()
+			parsed_source = BeautifulSoup(source_code, "html.parser")
+
+			emails.update(grabEmails(parsed_source.get_text())) # Grab emails
+			urlQueue.extend(grabLinks(parsed_source, domain))	# Grab urls and add to queue
+		except:
```

Comment:
I'm guarding against exceptions thrown by the mechanize library. Rather than crashing with a mechanize error, this program will just skip the link that brings about issues and continue checking other links in the queue. 

Such errors include invalid links (404) and requests disallowed by robots.txt (403), which occurs often when dealing with bigger sites like Google.